Fix the data inconsistency issue by moving the SetConsistentIndex into the transaction lock

ahrtr opened the pull request:
Fix [issues/13766 ](https://github.com/etcd-io/etcd/issues/13766).

Previously the `SetConsistentIndex()` is called during the apply workflow, but it's outside the backend transaction. Obviously they do not belong to an atomic operation any more. If a periodic commit happens between `SetConsistentIndex` and the left apply workflow, and etcd crashes for whatever reason right after the commit, then etcd runs into the data inconsistency issue. Please refer to discussion in [pull/13844](https://github.com/etcd-io/etcd/pull/13844) and [issues/13766](https://github.com/etcd-io/etcd/issues/13766).

In this PR, I moved the `SetConsistentIndex` into a `txPostLockHook`, which is executed each time right after the `batchTx.Lock()` being called. 

`batchTx.Lock()` is called in many places, but `txPostLockHook` is only supposed to be executed in the apply workflow. So I added one more interface method `LockWithoutHook` into `BatchTx`; it doesn't execute the `txPostLockHook` at all.
```
	// LockWithoutHook doesn't execute the txPostLockHook.
	LockWithoutHook()
```

Only the operations in the apply workflow call the `BatchTx.Lock`, and all others should call `BatchTx.LockWithoutHook`. If the operation doesn't have any impact on the apply workflow, such as the `etcdutl` commands, then it doesn't matter which lock it calls. 

cc all related people @ptabor @serathius @spzala  @wilsonwang371 @chaochn47 @tangcong @liuycsd @PaulFurtado @gyuho @hexfusion @jingyih @wpedrak Please take a look. Thanks. 

@chaochn47 @serathius @liuycsd @moonovo @michaljasionowski could you please try this PR to check whether you can still reproduce the data inconsistency issue? Thanks.

ptabor commented:
Thank you @ahrtr. The fix logic seems to me to be good. @serathius is the problem reproducible with the fix ?

What I'm afraid is that the solution is very error prone: 
  - it's easy to forget whether library-logic performing transaction should call `Unlock` vs `UnlockWithoutHooks`
  - it's difficult to asses whether the transitively called code is and will be calling the correct method. 

I'm think we need to have some method of verification that we don't forget or messed the calls. For example: 

If ETCD_VERIFY env variable (https://github.com/etcd-io/etcd/blob/0e83f62e0ca075268f5a34a7615129243bc8a52a/server/verify/verify.go#L30) is set 
      in Unlock & UnlockWithoutHooks we perform verification whether 'apply' method is/is-not (respectively) on the stack-trace. The mode is already enabled by default in both integrational & e2e tests. It would be slow... but it would catch most of the wrong calls... It might miss case when 'apply' is creating go-routine that performs a transaction... but I think it would be weird pattern we should be also aware off. 
     Alternative is to use 'ctx' for such verification... (implicit ctx driving business logic is imho anti-pattern) but it would require us to pass `ctx` to all methods that performs transactions (that seems a good thing in general). 

I reserved a ~full day for tomorrow to potentially try hands-on some of the ideas proposed above.

ahrtr commented:
Just added detailed comment to explain the difference between Lock and LockWithoutHook. @ptabor

serathius commented on server/storage/backend/batch_tx.go:
```diff
@@ -53,6 +53,15 @@ type BatchTx interface {
+	// workflow, and all other operations should call LockWithoutHook(). If the operation
+	// doesn't have any impact on the applying workflow, such as the `etcdutl` commands,
+	// then it doesn't matter which lock method it calls.
+	LockWithoutHook()
```
Having the preferred locking method called `LockWithoutHooks` is just asking for trouble. Based on the comment looks like `LockWithoutHooks` should be used in most of the code and `Lock` only in one place apply logic. 

Can we find a way to split transaction interface, so that `Lock` in apply would use hooks and `Lock` everywhere else will not use hooks?

ahrtr replied:
Thanks @serathius for the feedback. It's a valid comment, and @ptabor raised similar concern above. 

Firstly, if you go through the apply workflow, you will find that the users of `Lock` in the apply workflow overnumbers the users of `LockWithoutHooks` in other places.

Secondly, I have been thinking all the possible solutions (see below) in the past week, and eventually the solution in this PR is the simplest one, accordingly the least impact. 

All the possible solutions I can think of:
1. add a LockWithoutHook, which is supposed to be called outside the apply workflow. It's exactly what this PR delivers. It's the simplest solution and the least impact. The bad side is it's error prone, because it may be a little difficult to main the correct calls just as @ptabor pointed out.
2. add a `backend` wrapper (originally proposed by @ptabor ) and a `batchTxBuffered` wrapper; it just adds an additional functionality hooking on top of the existing design/implementation. Only the apply workflow holds a reference to the wrapped backend, all others keep unchanged.  It's the most decent solution so far. But the bad side is it needs big refactor the repo. See the concept in [pull/17](https://github.com/ahrtr/etcd/pull/17)
3. to be done (need to have breakfast for now)

michaljasionowski commented:
I still managed to reproduce inconsistencies due to downgrade/upgrade as described in https://github.com/etcd-io/etcd/issues/13514#issuecomment-1080266201 with https://github.com/etcd-io/etcd/pull/13844 cherry-picked to v3.5.2. I didn't try this PR, as it's more complex to cherry-pick.

deads2k commented:
With this PR, the test from https://github.com/etcd-io/etcd/pull/13838 no longer produces "found data inconsistency with peers"

ahrtr commented:
> With this PR, the test from #13838 no longer produces "found data inconsistency with peers"

Thanks @deads2k for the good news. Just to double confirm, can you easily reproduce this issue without this PR in your environment?

serathius commented:
I have also run qualification and was not longer able to reproduce the issue on this PR. To give some data I have run qualification 3 times each taking 10 minutes and didn't find any issues. Compare this to reproduction on `main` branch, which I run 5 times each of them managing to reproduce the issue on average in 2 minutes (minimal 30 seconds, maximum 9 minutes)

dims commented:
this is great news @serathius Nice hob @ahrtr !

deads2k commented:
> Thanks @deads2k for the good news. Just to double confirm, can you easily reproduce this issue without this PR in your environment?

Yes, using https://github.com/etcd-io/etcd/pull/13838 without this PR, I'm able to produce the  "found data inconsistency with peers" failure reliably.

ptabor commented:
Still looking into this.  Added verification as described above and discovering subtle cases as expected: 

1. We do call (read-only): Authentication from the applier. 
    This illustrate that we need to assume there are idempotent/reentrant transaction that will NOT move consistency_index forward. So it's OK, but we cannot assume all nested-apply transaction will call .Lock(WithHook). 
Alternatively `authBackendTx` needs to support `RLock`.
```
go.etcd.io/etcd/server/v3/storage/backend.(*batchTx).LockWithoutHook
	etcd/server/storage/backend/batch_tx.go:106
go.etcd.io/etcd/server/v3/storage/schema.(*authBatchTx).LockWithoutHook
	etcd/server/storage/schema/auth.go:112
go.etcd.io/etcd/server/v3/storage/schema.(*authBackend).GetUser
	etcd/server/storage/schema/auth_users.go:24
go.etcd.io/etcd/server/v3/auth.(*authStore).Authenticate
	etcd/server/auth/store.go:315
go.etcd.io/etcd/server/v3/etcdserver.(*applierV3backend).Authenticate
	etcd/server/etcdserver/apply.go:827
go.etcd.io/etcd/server/v3/etcdserver.(*applierV3backend).Apply
	etcd/server/etcdserver/apply.go:197
go.etcd.io/etcd/server/v3/etcdserver.(*authApplierV3).Apply
	etcd/server/etcdserver/apply_auth.go:61
go.etcd.io/etcd/server/v3/etcdserver.(*EtcdServer).applyEntryNormal
	etcd/server/etcdserver/server.go:1870
go.etcd.io/etcd/server/v3/etcdserver.(*EtcdServer).apply
	etcd/server/etcdserver/server.go:1777
go.etcd.io/etcd/server/v3/etcdserver.(*EtcdServer).applyEntries
	etcd/server/etcdserver/server.go:1084
go.etcd.io/etcd/server/v3/etcdserver.(*EtcdServer).applyAll
	etcd/server/etcdserver/server.go:904
go.etcd.io/etcd/server/v3/etcdserver.(*EtcdServer).run.func8
	etcd/server/etcdserver/server.go:836
go.etcd.io/etcd/pkg/v3/schedule.(*fifo).run
	etcd/pkg/schedule/schedule.go:157
```

And the other way around:  Authenticate is being called outside of 'apply' from interceptors... 
```
go.etcd.io/etcd/server/v3/storage/backend.(*batchTx).Lock
	/etcd/server/storage/backend/batch_tx.go:89
go.etcd.io/etcd/server/v3/storage/schema.(*authBatchTx).Lock
	/etcd/server/storage/schema/auth.go:108
go.etcd.io/etcd/server/v3/auth.(*authStore).CheckPassword.func1
	/etcd/server/auth/store.go:350
go.etcd.io/etcd/server/v3/auth.(*authStore).CheckPassword
	/etcd/server/auth/store.go:363
go.etcd.io/etcd/server/v3/etcdserver.(*EtcdServer).Authenticate
	/etcd/server/etcdserver/v3_server.go:434
go.etcd.io/etcd/server/v3/etcdserver/api/v3rpc.(*AuthServer).Authenticate
	/etcd/server/etcdserver/api/v3rpc/auth.go:57
go.etcd.io/etcd/api/v3/etcdserverpb._Auth_Authenticate_Handler.func1
	/etcd/api/etcdserverpb/rpc.pb.go:8064
go.etcd.io/etcd/pkg/v3/grpc_testing.(*GrpcRecorder).UnaryInterceptor.func1
	/etcd/pkg/grpc_testing/recorder.go:38
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
	/Users/ptab/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25
github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1
	/Users/ptab/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
	/Users/ptab/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25
go.etcd.io/etcd/server/v3/etcdserver/api/v3rpc.newUnaryInterceptor.func1
	/etcd/server/etcdserver/api/v3rpc/interceptor.go:74
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
	/Users/ptab/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25
go.etcd.io/etcd/server/v3/etcdserver/api/v3rpc.newLogUnaryInterceptor.func1
	/etcd/server/etcdserver/api/v3rpc/interceptor.go:81
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
	/Users/ptab/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1
	/Users/ptab/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:34
go.etcd.io/etcd/api/v3/etcdserverpb._Auth_Authenticate_Handler
	/etcd/api/etcdserverpb/rpc.pb.go:8066
google.golang.org/grpc.(*Server).processUnaryRPC
	/Users/ptab/go/pkg/mod/google.golang.org/grpc@v1.41.0/server.go:1279
google.golang.org/grpc.(*Server).handleStream
	/Users/ptab/go/pkg/mod/google.golang.org/grpc@v1.41.0/server.go:1608
google.golang.org/grpc.(*Server).serveStreams.func1.2
	/Users/ptab/go/pkg/mod/google.golang.org/grpc@v1.41.0/server.go:923
```

ahrtr commented:
Thanks @ptabor . 

In your first stack trace, we do need to call `Lock()`. It's coming from the apply workflow, so we need to move the `consistent_index` forward although it's read-only.

In your second stack trace coming from outside `apply`, yes. It's a good catch. 

I just enhanced `authBackend` to support read-only ReadTx. Please note that some read-only operations may also need to move the consistent_index forward if they are coming from the apply workflow. But the [GetUser](https://github.com/etcd-io/etcd/blob/main/server/storage/schema/auth_users.go#L22) is a corner case, because it's called by a couple of places; some calling are coming from the apply workflow, while others are not. So I addressed it separately. 

I rebased this PR, so all the three commits are submitted again, but you only need to take a look at the last one to check the latest change I made.

ahrtr commented:
I enabled the `EXPECT_DEBUG`, realized that the cov test is always failed. Steps to reproduce it:

```
$ export EXPECT_DEBUG=true
$ cd tests/e2e
$ go test --tags='cov cluster_proxy' -run TestClusterVersion/When_start_servers_one_by_one
```

The error message: 
```
/Users/wachao/go/src/github.com/ahrtr/etcd/bin/etcd_test-29228: {"level":"fatal","ts":"2022-04-04T15:03:23.206+0800","caller":"verify/verify.go:104","msg":"Verification failed","data-dir":"/var/folders/n9/qhsds19d75x16c351t6tzhfc0000gp/T/TestClusterVersionWhen_start_servers_one_by_one1734509272/004","error":"backend.ConsistentIndex (9) must be >= last snapshot index (11)","stacktrace":"go.etcd.io/etcd/server/v3/verify.MustVerifyIfEnabled\n\t/Users/wachao/go/src/github.com/ahrtr/etcd/server/verify/verify.go:104\ngo.etcd.io/etcd/server/v3/embed.(*Etcd).Close.func1\n\t/Users/wachao/go/src/github.com/ahrtr/etcd/server/embed/etcd.go:386\ngo.etcd.io/etcd/server/v3/embed.(*Etcd).Close\n\t/Users/wachao/go/src/github.com/ahrtr/etcd/server/embed/etcd.go:446\ngo.etcd.io/etcd/pkg/v3/osutil.HandleInterrupts.func1\n\t/Users/wachao/go/src/github.com/ahrtr/etcd/pkg/osutil/interrupt_unix.go:68"}
```


The consistent_index is indeed 9, 
```

$ ./etcd-dump-db iterate-bucket /Users/wachao/tmp/etcd/test_pr/t2 meta --decode
key="term", value=2
key="storageVersion", value="3.6.0"
key="consistent_index", value=9
key="confState", value="{\"voters\":[1632129655199559124,8740191053076832019,17148105919378895959],\"auto_leave\":false}"
```

The index in snap is indeed 11, and there is a panic when parsing the logs using the tool,
```

$ ./etcd-dump-logs /Users/wachao/tmp/etcd/test_pr/t2
Snapshot:
term=4 index=11 nodes=[16a67d46258cedd4 794b65748bb11f13 edfa4f7aa75a3457] confstate={"voters":[1632129655199559124,8740191053076832019,17148105919378895959],"auto_leave":false}
Start dumping log entries from snapshot.
WAL metadata:
nodeID=794b65748bb11f13 clusterID=38567eb454ea0afa term=4 commitIndex=11 vote=794b65748bb11f13
WAL entries:
panic: runtime error: index out of range [-1]
 
goroutine 1 [running]:
main.main()
    /Users/wachao/go/src/github.com/ahrtr/etcd/tools/etcd-dump-logs/main.go:127 +0xf9a
```

Will continue to investigate it.

ahrtr commented:
Eventually I figured out the root cause for the pipeline failure of coverage test, please see my last commit in this PR.

serathius commented on server/auth/store.go:
```diff
@@ -935,6 +939,8 @@ func NewAuthStore(lg *zap.Logger, be AuthBackend, tp TokenProvider, bcryptCost i
 
 	be.CreateAuthBuckets()
 	tx := be.BatchTx()
+	// We should call LockWithoutHook here, but the txPostLockHoos isn't set
```
I would worry about introducing this assumption. It might be true now, but we should minimize number of logical dependencies that cannot be enforced in code.

ahrtr replied:
All operations outside the apply workflow are supposed to call `LockWithoutHook`, so does `NewAuthStore` . But it's still in progress of etcd startup/initialization (because it's called in [NewServer](https://github.com/etcd-io/etcd/blob/a08d479463960948d5d9c3508623ade915c6e66f/server/etcdserver/server.go#L373)), which means etcd hasn't started to serve any requests yet, so it definitely has no any impact on the apply workflow. 

From the implementation perspective, we don't even have an instance of [EtcdServer](https://github.com/etcd-io/etcd/blob/a08d479463960948d5d9c3508623ade915c6e66f/server/etcdserver/server.go#L202) yet, it isn't possible to call the [getTxPostLockHook](https://github.com/ahrtr/etcd/blob/2c961171cd07f7fba2da9fe41ca8507a81b9ec1b/server/etcdserver/server.go#L2332) 

So we should be good here.

serathius replied:
PTAL https://github.com/etcd-io/etcd/issues/13883
Looks like cache in AuthStore is not reset correctly. Depending on implementation of fix we might decide to call Recover during etcd runtime

ahrtr replied:
Please note that the logic (auth.Recover) is already there, see [server.go#L1012](https://github.com/etcd-io/etcd/blob/a08d479463960948d5d9c3508623ade915c6e66f/server/etcdserver/server.go#L1012). The issue [13883](https://github.com/etcd-io/etcd/issues/13883) is just related to `rangePermCache`, so it should have no any impact on this PR.

ahrtr commented:
The test is green now, we can ignore the `semaphoreci` failure, which is caused by the dockerhub rate limit.

```

Step 1/14 : FROM ubuntu:21.10
toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
```

serathius commented:
Talked with @ptabor about how we should proceed with the fix. During review we both raised the issue of the fix being error prone. It splits the current locking method into two (`Lock` and `LockWithoutHook`) and requires us to manually pick one of them. Which method to pick depends on stacktrace, whether call was done as part of apply workflow or not. Fixing the data inconsistency requires us to correctly pick locking method in tens of cases. We make any mistake and the issue will remain, so we don't think this is acceptable.

Piotr spent some time checking if we can refactor the apply code in a way that would automatically make the decision if lock will use hooks https://github.com/etcd-io/etcd/pull/13878. This refactor ended up too big to be acceptable for v3.5 backport. We will revisit it after the immediate issue is fixed.
 
 Piotr also looked into implementing a automatic verification whether `Lock` or `LockWithoutHooks` were correctly picked. If `ETCD_VERIFY` environment variable is passed, locking will automatically check the stacktrace and panic if incorrect method was picked. Enabling this on all etcd seems like a best way to ensure that fix is correct. 
 
We agree that there are two remaining issues that we want addressed before we merge this PR:
* Automatic verification is enabled on all etcd tests.
* Names of functions are improved to imply when they should be used. One proposal `Lock` (with hook) should be renamed to `LockWithinApply` and  `LockWithoutHook` is renamed to `LockOutsideApply`.

Piotr will be busy for the rest of the week, so I will look into implementing automatic verification. I will send minimal PR asap.

@ahrtr can you do the function rename?

ahrtr commented:
Thanks @serathius and @ptabor for the feedback, which makes sense to me!

`LockWithinApply` and `LockOutsideApply` look better. The only downside is that the `backend` shouldn't care about the business logic (the apply workflow). But I agree they are clearer, so I will rename the function asap (will finish it tomorrow). 

Just as raised in [discussion_r840127417](https://github.com/etcd-io/etcd/pull/13854#discussion_r840127417), I tried to refactor the project to wrap backend, eventually I realize that it's a huge change, and definitely not acceptable for 3.5.

ahrtr commented:
How about `LockInsideApply` and `LockOutsideApply`?

serathius commented:
PR with verification https://github.com/etcd-io/etcd/pull/13885

ptabor commented:
FTR: I exported the PR: https://github.com/etcd-io/etcd/pull/13884/commits/08c9689e9d9425df6a4f32f5c066eeccdea1ad2b where I was prototyping the verification.

Lessons learned: 
  - It's good to use refactoring tools to automatically translate `Lock` -> `LockOutsideApply`. The unit tests are full of these locks, that will keep complaining. 
  - Use  ``` func applyEntries(x func()){ x() } ``` in tests to wrap the unit-code that executes the 'apply' style code
  - Tests that call `NewDefaultBackend`:  are failing the verification silently due to `zap.Nop()` logger being used:
     https://github.com/etcd-io/etcd/pull/13884/commits/02149bff1d74b3c36763fc97d8197170873ba449
 - The explilicit `RLock` for e.g. Auth seems like the way to go.  Thank you for adding it.

serathius commented:
Sent https://github.com/etcd-io/etcd/pull/13886 to address no logger passed to backend.

ahrtr commented:
All comments are resolved and all tests are green now. 

Please see the following diagram. BatchTx exposes three lock interface methods below,
1. Lock
2. LockInsideApply
3. LockOutsideApply

![batchtx_lock](https://user-images.githubusercontent.com/30739825/161902290-c396cb30-399c-4b97-b610-d4125b7e1d2a.png)

Note that all unit tests can keep unchanged, because they can still call `Lock` directly. Of course, unit tests can also call `LockInsideApply` or `LockOutsideApply` if we intentionally add test cases to test them, such as `verify_test.go`. But if the callers of `Lock` are not coming from unit tests, then you will see a panic, see [batch_tx.go#L73](https://github.com/ahrtr/etcd/blob/fea4bf55e1fc03e28e1ac573722fb5a9b9abf54e/server/storage/backend/batch_tx.go#L73). I intentionally added this check to avoid misuse of the Lock, and in the meanwhile minimize the effort of updating the existing unit test. 

The BatchTx interface has a nested ReadTx, in which Lock/Unlock/RLock/RUnlock are defined, so BatchTx must implement the all interface methods, of course including Lock. There are a couple of todo items:
1. It doesn't comply with ISP (Interface Segregate Principle), BatchTx shouldn't know/implement RLock/RUnlock at all. 
2. Even we decouple BatchTx and ReadTx in future, and remove the `Lock` from `batchTx` (so that users must call `LockInsideApply` or `LockOutsideApply`), users can still call `Lock()` directly, because it has a nested field [sync.Mutex](https://github.com/etcd-io/etcd/blob/ad03f2076ad418dcbf4b6976665819448b182f49/server/storage/backend/batch_tx.go#L61), accordingly *batchTx includes promoted methods with receiver `*sync.Mutex`. So If we want to completely remove `Lock` in order to avoid any confusion, then we can change the nested field `sync.Mutex `to a normal field, and of course, we also need to update all existing unit tests which use `Lock`.

I think we should be all good so far.

ahrtr commented on server/storage/backend/hooks_test.go:
```diff
@@ -41,8 +41,6 @@ func TestBackendPreCommitHook(t *testing.T) {
 	// Empty commit.
 	tx.Commit()
 
-	write(tx, []byte("foo"), []byte("bar"))
```
In case someone asks why I removed this line...  it causes instability, because the periodical commit goroutine may contribute one more commit during the test. The symptom is the test may occasionally fail, see error message below,
```

$ go test -run TestBackendPreCommitHook
--- FAIL: TestBackendPreCommitHook (0.29s)
    hooks_test.go:120: OnPreCommit executed: key >
    hooks_test.go:120: OnPreCommit executed: key >c
    hooks_test.go:120: OnPreCommit executed: key >cc
    hooks_test.go:46: 
        	Error Trace:	hooks_test.go:46
        	Error:      	Not equal: 
        	            	expected: ">cc"
        	            	actual  : ">ccc"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	->cc
        	            	+>ccc
        	Test:       	TestBackendPreCommitHook
        	Messages:   	expected 2 explict commits
    hooks_test.go:120: OnPreCommit executed: key >ccc
    hooks_test.go:48: 
        	Error Trace:	hooks_test.go:48
        	Error:      	Not equal: 
        	            	expected: ">ccc"
        	            	actual  : ">cccc"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	->ccc
        	            	+>cccc
        	Test:       	TestBackendPreCommitHook
        	Messages:   	expected 3 explict commits
    hooks_test.go:120: OnPreCommit executed: key >cccc
FAIL
exit status 1
FAIL	go.etcd.io/etcd/server/v3/storage/backend	0.750s
```

serathius commented on server/auth/store.go:
```diff
@@ -935,6 +939,8 @@ func NewAuthStore(lg *zap.Logger, be AuthBackend, tp TokenProvider, bcryptCost i
 
 	be.CreateAuthBuckets()
 	tx := be.BatchTx()
+	// We should call LockOutsideApply here, but the txPostLockHoos isn't set
```
Let's just call `LockOutsideApply`

ahrtr replied:
It's authBatchTx instead of the batchTx, see [auth.go#L108](https://github.com/ahrtr/etcd/blob/4d1070af9e835b5dc7bb6f5924f663a5b2c10973/server/storage/schema/auth.go#L108).

ahrtr replied:
Majority calls of` (atx *authBatchTx) Lock() ` are coming from the apply workflow, but this one is a special case. 

Please note that I enhanced the authBackend to support ReadTx (previously it only supports BatchTx), and all readonly operations go to the ReadTx, all writing operations go to the BatchTx.

serathius commented on server/etcdserver/cindex/cindex.go:
```diff
@@ -73,7 +77,19 @@ func (ci *consistentIndex) ConsistentIndex() uint64 {
+	return v
+}
+
+// UnsafeConsistentIndex is similar to ConsistentIndex,
```
This comment is not needed as `Unsafe` is very common convention in etcd code.

ahrtr replied:
done.

ahrtr commented on etcdutl/etcdutl/backup_command.go:
```diff
@@ -322,7 +322,7 @@ func saveDB(lg *zap.Logger, destDB, srcDB string, idx uint64, term uint64, desir
 	if !v3 {
 		tx := be.BatchTx()
-		tx.Lock()
+		tx.LockInsideApply()
```
Actually it doesn't matter because the txPostLockHook is nil, but I agree LockOutsideApply is clearer. Done.

serathius commented on server/storage/schema/auth_roles.go:
```diff
@@ -32,6 +32,40 @@ func (abe *authBackend) GetRole(roleName string) *authpb.Role {
 }
 
 func (atx *authBatchTx) UnsafeGetRole(roleName string) *authpb.Role {
```
Any reason to reorder functions? It might make backport harder.

ahrtr replied:
I did not reorder the methods, instead i just added some new methods. It's just a github displaying issue.

serathius replied:
ack

ptabor commented on server/etcdserver/server.go:
```diff
@@ -207,8 +207,10 @@ type EtcdServer struct {
 
-	consistIndex cindex.ConsistentIndexer // consistIndex is used to get/set/save consistentIndex
-	r            raftNode                 // uses 64-bit atomics; keep 64-bit aligned.
+	consistentIdx  uint64                   // must use atomic operations to access; keep 64-bit aligned.
```
Why we are adding this instead of use internals of 'consistIndex' ?

ahrtr replied:
I just added `consistentIdx` and `consistentTerm` to temporarily save the e.Index and e.Term, and in the txPostLockHook they are saved to the ConsistentIndex and Term by calling `s.consistIndex.SetConsistentIndex(cIdx, term)`. The original `consistIndex` keeps unchanged. 

```

	consistentIdx  uint64                   // must use atomic operations to access; keep 64-bit aligned.
	consistentTerm uint64                   // must use atomic operations to access; keep 64-bit aligned.
	consistIndex   cindex.ConsistentIndexer // consistIndex is used to get/set/save consistentIndex
```

ptabor replied:
I still don't understand.  cindex.ConstentIndexer has 2 purposes: 
  a) storing the cindex temporarilly `(consistentIndex, term)` variables. 
  b) Flushing it to the backend with 'UnsafeSave' is called. 

Why the hook cannot use UnsafeSave to flush it and  `consistIndex.SetConsistentIndex()` as the temporary storage ?

ahrtr replied:
Previously, we save the consistent_index and term into cindex by calling s.consistIndex.SetConsistentIndex(e.Index, e.Term) in applyEntries, and persists the values into db in preCommitHook. So the data path is:
```
e.Index, e.Term --> ci.consistentIndex, ci.term  --> db
```

Now, I added one more layer of cache, because we can't save the `e.Index and e.Term` into `ci.consistentIndex and ci.term ` directly in applyEntries, 
```
e.Index, e.Term --> s.consistentIdx, s.consistentTerm -->  ci.consistentIndex, ci.term --> db
```

ptabor replied:
> we can't save the e.Index and e.Term into ci.consistentIndex and ci.term 

Solution I'm thinking about is following: 
 1. We remove the call from the hook: https://github.com/etcd-io/etcd/blob/d24ef3ac20ebfb7fa1f7c2c1a693c5c80ffcb136/server/storage/hooks.go#L44-L45
2. We call it directly from (SetTxPostLockHook)

I might be missing something obvious, but I think that: 
 1) We are getting simpler code without another `s.consistentIdx` field exposed that one can use as authoritative state of the raft. The state exposed in EtcdServer and not encapsulated in some object/interface with clear purpose is likely to be misused as the class is widely provided within the huge package. 
 2) We might loose some performance, because we will be setting the `key` in a single bolt transaction multiple times. I don't think it's a significant performance gain, as (I think) bolt in scope of a single transaction is working in-place on the same page. 

I think that 1) is more important than 2).  Is there another  motivation I'm missing ?
If there is ... we should at least pack the SetTxPostLockHook and the fields in a single object that encapsulates the logic.

ahrtr replied:
Yes, we are on the same page now. Just as I mentioned in the third proposal in [discussion_r840127417](https://github.com/etcd-io/etcd/pull/13854#discussion_r840127417), it doesn't make much sense to persist the consistent_index in each apply (not the bbolt db transaction), because it will cause decrease of performance. If we follow this way, then what's value of `OnPreCommitUnsafe`?

I agree that introducing s.consistentIdx into Etcdserver is not a good way of encapsulation.  Probably we can move the two fields into [consistentIndex](https://github.com/etcd-io/etcd/blob/d24ef3ac20ebfb7fa1f7c2c1a693c5c80ffcb136/server/etcdserver/cindex/cindex.go#L47). WDYT?

ptabor replied:
consistentIndex or new object seems good to me. 
 we can call them: 
     "applyingIndex" and "applyingTerm" for example to distingush from "appliedIndex & appliedTerm". 
     
 I think that at the time the code was written... it seemed as not difference/cheap optimisation to delay the write to "on commit". There was no measurement whether it's actual improvement.

ahrtr replied:
Yes, after second thought I'd agree that there shouldn't much performance gain by introducing the `OnPreCommitUnsafe`, because the bbolt DB loads a page to a node only once in each transaction, and any following updates are only in the node (in memory).  The only gain is reducing multiple operations on node to only one. But I agree we need to run some benchmark to compare the performance difference. 

I will take a look at this once this issue is resolved. Afterwards, we can make a decision whether or not we should remove the `OnPreCommitUnsafe`.

But for now, let's just keep it as it's. 

`"applyingIndex" and "applyingTerm"` make perfect sense to me. Will move them into `consistentIndex`. Thanks.

ahrtr replied:
Thanks & Done.

ahrtr replied:
Please take a look at the last commit in this PR.

ahrtr commented on etcdutl/etcdutl/backup_command.go:
```diff
@@ -322,7 +322,7 @@ func saveDB(lg *zap.Logger, destDB, srcDB string, idx uint64, term uint64, desir
 	if !v3 {
 		tx := be.BatchTx()
-		tx.Lock()
+		tx.LockInsideApply()
```
I have already changed it to `LockOutsideApply`.  Actually it doesn't matter which method it calls, because the txPostLockHook is nil, see [batch_tx.go#L80](https://github.com/ahrtr/etcd/blob/8259df98235f9887fe022e583bd56e71d88a5e6c/server/storage/backend/batch_tx.go#L80)

ahrtr commented on server/storage/backend/backend.go:
```diff
@@ -67,6 +67,9 @@ type Backend interface {
 	Close() error
+
+	// SetTxPostLockHook sets a txPostLockHook.
+	SetTxPostLockHook(func())
```
Actually it should be PreApply, because the hook is executed right after requiring the lock.

ahrtr replied:
I'd like to keep the current name. The concept is clear, currently we have `OnPostLock` (added in this PR). Probably we can support `OnPreUnlock` in future if needed. 

@ptabor  do you have strong opinion for the renaming?

ahrtr commented on server/storage/backend/batch_tx.go:
```diff
@@ -65,18 +65,31 @@ type batchTx struct {
 }
 
+// Lock is supposed to be called only by the unit test.
 func (t *batchTx) Lock() {
```
Yes, it's a good question, which I have already answered in [issuecomment-1089856121](https://github.com/etcd-io/etcd/pull/13854#issuecomment-1089856121).

ptabor replied:
Please leave a TODO to clean it up later, referencing the https://github.com/etcd-io/etcd/pull/13854#issuecomment-1089856121 comment and I'm OK with closing this thread.

ahrtr replied:
There is already one [TODO item ](https://github.com/etcd-io/etcd/blob/main/server/storage/backend/batch_tx.go#L89-L91 ). I will take care of it later.

ahrtr commented on server/storage/backend/verify.go:
```diff
@@ -54,3 +63,8 @@ func insideApply() bool {
+
+func fromUnittest() bool {
+	stackTraceStr := string(debug.Stack())
+	return strings.Contains(stackTraceStr, "_test.go")
```
Yes, but I do not see any cases which call `Lock` directly so far. And it's also OK for integration to call it.

serathius replied:
Should we add `&& !strings.Contains(stackTraceStr, "tests/")` to avoid this returning true for integration/e2e tests?

ahrtr replied:
Good idea. Thanks & resolved.

ahrtr commented on server/storage/schema/schema.go:
```diff
@@ -31,7 +31,7 @@ var (
-	tx.Lock()
+	tx.LockOutsideApply()
 	defer tx.Unlock()
 	return unsafeValidate(lg, tx)
```
Yes, you are right. Updated.

serathius commented on server/storage/backend/verify.go:
```diff
@@ -46,6 +46,15 @@ func ValidateCalledOutSideApply(lg *zap.Logger) {
 	}
 }
 
+func ValidateCalledFromUnittest(lg *zap.Logger) {
```
```suggestion
func ValidateCalledInsideUnittest(lg *zap.Logger) {
```
Let's be consistent in naming functions here.

serathius replied:
Either way is fine, however let's have a one way to express stacktrace analysis.

ahrtr replied:
done.

serathius commented on server/storage/backend/verify.go:
```diff
@@ -46,6 +46,15 @@ func ValidateCalledOutSideApply(lg *zap.Logger) {
+	if !verifyLockEnabled() {
+		return
+	}
+	if !fromUnittest() {
```
```suggestion
	if !insideUnittest() {
```

serathius commented on server/storage/backend/verify.go:
```diff
@@ -54,3 +63,8 @@ func insideApply() bool {
 	return strings.Contains(stackTraceStr, ".applyEntries")
 }
+
+func fromUnittest() bool {
```
```suggestion
func insideUnittest() bool {
```

ahrtr replied:
makes sense, updated.

serathius commented on server/storage/backend/batch_tx.go:
```diff
@@ -65,18 +65,31 @@ type batchTx struct {
 
+// Lock is supposed to be called only by the unit test.
 func (t *batchTx) Lock() {
+	ValidateCalledFromUnittest(t.backend.lg)
```
```suggestion
	ValidateCalledInsideUnittest(t.backend.lg)
```

ahrtr replied:
makes sense, updated.

ptabor commented on server/etcdserver/cindex/cindex.go:
```diff
@@ -106,18 +148,24 @@ type fakeConsistentIndex struct {
 }
 
-func (f *fakeConsistentIndex) ConsistentIndex() uint64 { return f.index }
+func (f *fakeConsistentIndex) ConsistentIndex() uint64                   { return f.index }
```
nit: If we are storing as atomic, we should follow the pattern and load (`atomic.LoadUint64`) as atomic as well.

ahrtr replied:
thanks & done.

ptabor commented on server/storage/backend/backend.go:
```diff
@@ -119,6 +122,9 @@ type backend struct {
 	hooks Hooks
 
+	// txPostLockHook is called each time right after locking the tx.
+	txPostLockHook func()
```
txPostLockInsideApplyHook

ahrtr replied:
makes sense, updated.

ptabor reviewed (approved):
Thank you so much !!!
Great work.

serathius commented on server/etcdserver/cindex/cindex.go:
```diff
@@ -58,6 +64,19 @@ type consistentIndex struct {
+	// and raftpb.Entry.Term, and they are not ready to be persisted yet. They will be
+	// saved to consistentIndex and term above in the txPostLockInsideApplyHook.
+	//
+	// TODO(ahrtr): try to remove the OnPreCommitUnsafe, and compare the
```
Can we do the performance check before merging this PR? I would prefer not to backport a PR with todo. Please let me know if you need help with measuring it.

ahrtr replied:
I think we are all good so far. I will investigate whether or not we should completely remove the `OnPreCommitUnsafe` in the next step **only** in the **main** branch.  The change will not be cherry picked to 3.5. 

We can use the tools/benchmark, correct?

serathius replied:
I wanted to backport this change because it makes the code much simpler, which will be important for long term maintenance of release-3.5 branch. Just making sure that the decision not to fix it imminently is not rushed, let's not sacrifice the quality, however it you think it's not worth backporting that's also ok.

ahrtr replied:
Your comment is valid, but usually we only backport bug fixes to previous release, so does this PR. But completely possibly removing `OnPreCommitUnsafe` is a refactor, and so I don't think we should backport it to 3.5.  Please note that refactoring may also introduce regression, so we should only do it in main branch.

ptabor replied:
I think this PR is good to go.

serathius commented on server/etcdserver/server.go:
```diff
@@ -401,6 +401,10 @@ func NewServer(cfg config.ServerConfig) (srv *EtcdServer, err error) {
 
+	// Set the hook after EtcdServer finishes the initialization to avoid
+	// the hook being called during the initialization process.
+	srv.be.SetTxPostLockInsideApplyHook(srv.getTxPostLockInsideApplyHook())
```
I'm thinking if there is a better place to install the hook. Putting it on pile with other EtcdServer initialization doesn't seem correct. My thought was that this hook is for CI and should be owned by CI. Have you considered moving the hook installation into `(ci *consistentIndex) SetBackend`?

ahrtr replied:
Yes, I thought about moving it into consistentIndex, and it looks more reasonable and clearer. But the reason why I gave up is that we need to make sure the hook is only set after Etcdserver finishes being initialized. The authBatchTx may call LockInsideApply during Etcdserver initialization, the call path is [NewAuthStore](https://github.com/ahrtr/etcd/blob/data_corruption/server/auth/store.go#L944)  --> [authBatchTx.Lock](https://github.com/ahrtr/etcd/blob/4033f5c2b9b57cd341d1b3522abb03bedbeb9a07/server/storage/schema/auth.go#L109) -- > `BatchTx.LockInsideApply`

serathius commented:
Can you squash the commits?

ahrtr commented:
> Can you squash the commits?

I intentionally added multiple commits, each has a clear goal/comment. I can merge the first two commits. WDYT?

serathius commented:
@ahrtr can you handle backport to v3.5?

ahrtr commented:
> @ahrtr can you handle backport to v3.5?

Definitely yes, will do it today.  

Many thanks to  @ptabor and @serathius for all the helps and valuable review comments.

ahrtr commented:
Also thanks to @deads2k @serathius and @michaljasionowski for the verification of issue based on this PR!

ahrtr commented:
FYI. In case anyone needs a simple [summary](https://github.com/ahrtr/etcd-issues/tree/master/issues/13766)
